### PR TITLE
feat: migrate rpc signing code to alloy

### LINKED ISSUE #4742: Migrate EIP-712 and signing code in `reth-rpc` to Alloy
### :warning: Base on #4737 (`alloy`) :warning:

Use `alloy-sol-types` for EIP-712 in `EthSigner`, re-implement `ethers_core::utils::hash_message` in `reth-primitives`:

1. https://github.com/paradigmxyz/reth/blob/675c6bfc390b17d274541ef251375741f48dd3c3/crates/rpc/rpc/src/eth/signer.rs#L89-L93
2. https://github.com/paradigmxyz/reth/blob/675c6bfc390b17d274541ef251375741f48dd3c3/crates/rpc/rpc/src/eth/api/sign.rs#L22-L29

### PR DESCRIPTION
**Pull request by supernovahs**
Closes #4742

### CODE REVIEW

**Review by onbjerg — APPROVED** [review]
looks good to me, pending @DaniPopes or @mattsse

**Review by mattsse — APPROVED** [review]
lgtm, but defer to @DaniPopes on how to merge this

**Review by DaniPopes — CHANGES_REQUESTED** [review]
@supernovahs Can you rebase to `alloy` (#4737) and change the target branch to it? Otherwise LGTM.

File: crates/primitives/src/lib.rs [hunk 1]
```diff
@@ -113,3 +113,45 @@ pub mod utils {
...
+    let len = message.len();
+    let len_string = len.to_string();
+
+    let mut eth_message = Vec::with_capacity(PREFIX.len() + len_string.len() + len);
+    eth_message.extend_from_slice(PREFIX.as_bytes());
+    eth_message.extend_from_slice(len_string.as_bytes());
+    eth_message.extend_from_slice(message);
+
+    H256(*keccak256(&eth_message))
+}
```
Comment (gakonst) [inline crates/primitives/src/lib.rs:153]: This feels like common code that we'd want to have as utilities in alloy core?

File: crates/primitives/src/lib.rs [hunk 2]
```diff
@@ -113,3 +113,45 @@ pub mod utils {
 pub mod kzg {
     pub use c_kzg::*;
 }
+
+/// Helpers for working with serde
+pub mod serde_helper;
```
Comment (gakonst) [inline crates/primitives/src/lib.rs:118]: where was this?

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/primitives/src/lib.rs [hunk 3]
```diff
@@ -113,3 +113,44 @@ pub mod utils {
...
+
+/// Returns the keccak256 hash for the given data.
+#[inline]
+pub fn keccak256(data: impl AsRef<[u8]>) -> H256 {
```
Comment (DaniPopes) [inline crates/primitives/src/lib.rs:122]: Use `alloy_primitives::keccak256`

File: crates/primitives/src/lib.rs (hunk 2 above)
Comment (DaniPopes) [reply crates/primitives/src/lib.rs:118]:
```suggestion
```

This was moved in the big list above

File: crates/primitives/src/lib.rs [hunk 4]
```diff
@@ -113,3 +113,44 @@ pub mod utils {
...
+    let message = message.as_ref();
+    let len = message.len();
+    let len_string = len.to_string();
+
+    let mut eth_message = Vec::with_capacity(PREFIX.len() + len_string.len() + len);
+    eth_message.extend_from_slice(PREFIX.as_bytes());
+    eth_message.extend_from_slice(len_string.as_bytes());
+    eth_message.extend_from_slice(message);
+
+    H256(*keccak256(&eth_message))
```
Comment (DaniPopes) [inline crates/primitives/src/lib.rs:152]:
```suggestion
    keccak256(&eth_message)
```

File: crates/primitives/src/lib.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/primitives/src/lib.rs:153]: I guess so, in `alloy_primitives::utils`

File: Cargo.toml [hunk 5]
```diff
@@ -104,13 +104,15 @@ revm = { git = "https://github.com/Evalir/revm/", branch = "reintroduce-alloy-re
...
-alloy-primitives = "0.3"
 alloy-rlp = "0.3"
 ethers-core = { version = "2.0", default-features = false }
 ethers-providers = { version = "2.0", default-features = false }
 ethers-signers = { version = "2.0", default-features = false }
 ethers-middleware = { version = "2.0", default-features = false }
 
+#alloy
+alloy-dyn-abi = {version = "0.3.2",default-features = false}
+alloy-primitives = {version = "0.3.2",default-features = false}
```
Comment (DaniPopes) [inline Cargo.toml:115]: Can you keep these under `eth` above?

File: crates/primitives/src/lib.rs [hunk 6]
```diff
@@ -113,3 +113,29 @@ pub mod utils {
...
+    let mut eth_message = Vec::with_capacity(PREFIX.len() + len_string.len() + len);
+    eth_message.extend_from_slice(PREFIX.as_bytes());
+    eth_message.extend_from_slice(len_string.as_bytes());
+    eth_message.extend_from_slice(message);
+
+    (keccak256(&eth_message))
+}
+
+#[cfg(any(test, feature = "arbitrary"))]
+pub use arbitrary;
```
Comment (DaniPopes) [inline crates/primitives/src/lib.rs:141]:
```suggestion
```

File: crates/primitives/src/lib.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/primitives/src/lib.rs]: We can merge this and upstream after, if you don't mind opening a PR to [alloy-rs/core](https://github.com/alloy-rs/core) @supernovahs after this PR?

File: Cargo.toml (hunk 5 above)
Comment (DaniPopes) [reply Cargo.toml:115]: Also you'll need to patch these with git version at the bottom of the file

**Comment by supernovahs** [comment]
Don't know what happened to my pc, I am seeing ethers-core errors :/  . Trying to fix .

**Comment by supernovahs** [comment]
Thanks for the help Dan

**Comment by supernovahs** [comment]
Integration test failed with no reason ?

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**